pkg/logstructured/logstructured.go

```diff
@@ -274,81 +357,65 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 
 	go func() {
 		defer wg.Done()
+		var lastRev int64
 		rev, events, err := l.log.List(ctx, "/", "", 1000, 0, false)
 		for len(events) > 0 {
 			if err != nil {
-				logrus.Errorf("failed to read old events for ttl")
+				logrus.Errorf("Failed to read old events for ttl, %v", err)
 				return
 			}
 
 			for _, event := range events {
 				if event.KV.Lease > 0 {
 					result <- event
 				}
+
+				if event.KV.ModRevision > lastRev {
+					lastRev = event.KV.ModRevision
+				}
 			}
 
 			_, events, err = l.log.List(ctx, "/", events[len(events)-1].KV.Key, 1000, rev, false)
 		}
+		lastListRevision <- lastRev
 	}()
 
 	go func() {
 		defer wg.Done()
-		for events := range l.log.Watch(ctx, "/") {
+		revision := <-lastListRevision
+		for events := range l.Watch(ctx, "/", revision) {
 			for _, event := range events {
 				if event.KV.Lease > 0 {
 					result <- event
 				}
 			}
 		}
-
 		logrus.Info("TTL events watch channel was closed")
 	}()
 
 	return result
 }
 
-func (l *LogStructured) ttl(ctx context.Context) {
-	// vary naive TTL support
-	queue := workqueue.NewDelayingQueue()
-	go func() {
-		for l.handleTTLEvents(ctx, queue) {
-		}
-	}()
-
-	for {
-		select {
-		case <-ctx.Done():
-			queue.ShutDown()
-			return
-		default:
-		}
-
-		for event := range l.ttlEvents(ctx) {
-			queue.AddAfter(event, time.Duration(event.KV.Lease)*time.Second)
-		}
-	}
-}
-
-func (l *LogStructured) handleTTLEvents(ctx context.Context, queue workqueue.DelayingInterface) bool {
-	item, shutdown := queue.Get()
-	if shutdown {
-		logrus.Info("TTL events work queue has shut down")
-		return false
+func loadTTLEventKV(ttlEventMap *sync.Map, key interface{}) *ttlEventKV {
+	value, ok := ttlEventMap.Load(key)
+	if !ok {
+		return nil
 	}
 
-	defer queue.Done(item)
-	event, ok := item.(*server.Event)
+	eventKV, ok := value.(*ttlEventKV)
 	if !ok {
-		return true
+		return nil
 	}
 
-	logrus.Tracef("TTL events key: %v, modRev: %v", event.KV.Key, event.KV.ModRevision)
+	return eventKV
+}
 
-	if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
-		logrus.Errorf("Failed to delete expired key: %v, requeueing", err)
-		queue.AddAfter(item, retryInterval)
-	}
-	return true
+func storeTTLEventKV(ttlEventMap *sync.Map, key string, modRevision int64, expiredAt time.Time) {
```

Comment:
make `storeTTLEventKV` take the whole KV, instead of getting the key, revision, and ttl for it.
make` storeTTLEventKV` return the TTL duration so that we can enqueue using it.